process: initial documentation of verification

pahmann opened the pull request:
This is the porting of the work done in the incubator project. 

It includes the following artifacts, adopted to the latest syntax and checks for S-Core:
- Guidance
- Work Products
- Report template
- Requirement to test linking
- Test case templates
- S-Core verification plan draft

Implements: #320

Also-by: @julian711

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+=====================
+
+This guideline outlines the responsibilities and procedures for developers performing verification activities (test
+case creation, inspection, and review) for S-Core documentation, Rust and C/C++ projects of the platform.
```
- should we leave out name "S-Core"?
- documentation + tooling? We also need some guidance to link our tools to process requirements for Tool Qualification

hoe-jo replied:
process requirements are still missing

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+
+* **Comprehensive Coverage:** Test cases should cover all functional and non-functional requirements, including
+  positive, negative, and boundary conditions. Specific attention should be given to corner cases and error handling.
+* **Unit Testing:** Focus on isolating and testing individual units or components of the code. Strive for high code coverage.
```
Shall we mention something like:
- Component Test -> Requiremements
- Unit Test -> Branch + Line Coveage?
- Also specify targets for "high code coverage"?

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+    * **Rust:** Utilize the built-in testing framework with ``#[test]`` attributes and the ``cargo test`` command.
+    * **Python:** Use ``pytest`` frameworks.
+    * **C++:** Use Google Test frameworks.
+* **Integration Testing:** Verify the interaction between different components or modules.
```
Mention ITF Framework?

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+    * **C++:** Use Google Test frameworks.
+* **Integration Testing:** Verify the interaction between different components or modules.
+* **Platform Testing:** Test the platform with configured features as a whole.
+* **Regression Testing:** Ensure that changes do not introduce new defects. Automate regression tests where possible.
```
Unclear, Regression testing will be done in CI?

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
For me this looks like more like the "concept description"
Currently I´m somehow missing the steps what I need to do when I want to start a unit test

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+     - valid | invalid
+     -
+     -
+   * - TestType
```
Provide more examples what this means

pahmann replied:
Test type descriptions like for boundary analysis is a topic which will be discussed within testing community (and documented in process).

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+     - Requirements-based test | Design based tests
+     - Assumptions of Use are treated as requirements.
+     -
+   * - DerivationTechnique
```
Provide more examples + description

hoe-jo commented on docs/process/process_areas/verification/_assets/requirements_workflow_verification.svg:
Maybe take out parent req part and test suite?

pahmann replied:
addressed in commit [ebc7e98](https://github.com/eclipse-score/score/pull/441/commits/ebc7e98debbb9ed7fffc358f40316ca99823c8c6)

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_report_template.rst:
We should try to setup a more automated template where more results get generated

pahmann replied:
The template already says "Note2: All the above lists are generated automatically". Do you require more?

aschemmel-tech commented on docs/platform_management_plan/software_verification.rst:
```diff
@@ -14,14 +14,372 @@
+guidance is given here: :doc:`/process/process_areas/verification/guidance/verification_guidance`. The identifier of the respective
+method is to be used as meta data (*TestType* and *DerivationTechnique*).
+
+.. list-table:: Software verification methods (TestType)
```
Findings from audit still to resolve, will be checked again in second iteration

aschemmel-tech reviewed (changes requested):
see inline comments, first iteration

pahmann commented:
@hoe-jo this commit in #441 is for you: [workproducts: update verification and requirements](https://github.com/eclipse-score/score/pull/441/commits/9127eaec546fa7b3ca6929309657c9694c2867c7)

I removed some duplicated items and added the requirements inspection to your work products, as you already have the template for it.

pahmann commented on docs/process/process_areas/verification/guidance/verification_report_template.rst:
```diff
@@ -0,0 +1,158 @@
+
+This document implements <add "need" link>
+
+   +---------------------------+-------------------------------+
```
Remove table and follow new guideline on document handling as in PR.

pahmann commented on docs/process/process_areas/verification/guidance/verification_report_template.rst:
```diff
@@ -0,0 +1,158 @@
+   +---------------------------+-------------------------------+
+   | AUTHOR                    | <Committer>                   |
+   +---------------------------+-------------------------------+
+   | Reviewer                  | <Quality Manager>             |
```
See above. Header will be replaced to match documentation management plan requirements

pahmann commented on docs/platform_management_plan/software_verification.rst:
```diff
@@ -14,14 +14,372 @@
+#. :need:`wp__verification__sw_unit_test`
+#. :need:`wp__sw_req_inspect`
+#. :need:`wp__sw_arch_verification`
+#. :need:`wp__sw_code_inspect`
```
It is a good point. Do we make it part of the wp__verification__sw_unit_test or should we have a dedicated wp for it?

pahmann replied:
I put it to sw_unit_test, but I am not really convinced. Maybe we can resolve this on Monday.

pahmann commented on docs/process/process_areas/verification/guidance/verification_process_toolreqs.rst:
```diff
@@ -0,0 +1,117 @@
+            {
+               RecordProperty("Status", "<Status>");
+               RecordProperty("ASIL", "<ASIL>");
+               RecordProperty("Priority", "<Priority>");
```
@hoe-jo  Do we really want to have Priority a property? Also it is not described anywhere so far and we want to execute all tests in CI. I wonder, if we really need it.

pahmann commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,178 @@
+------------------
+
+.. list-table:: Test specification properties
+    :header-rows: 1
```
changed to `:widths: 10 30 40 20`

aschemmel-tech commented on docs/process/process_areas/verification/workflows.rst:
```diff
@@ -0,0 +1,55 @@
+   :output: wp__verification__plan
+   :contains: gd_guidl__verification_guide
+
+   ``[TODO: Is this the right relationship for the verification guidance?]``
```
yes it is, if the gd_guidl__verification_guide explains how the Verification Plan is created.

aschemmel-tech commented on docs/process/process_areas/verification/workflows.rst:
```diff
@@ -0,0 +1,55 @@
+   as needed throughout the project lifecycle.
+
+
+``[TODO: Additional workflows to be added]``
```
resolve TODOs

pahmann replied:
Remaining ToDos which where spread around in the whole verification process are now all collected in issue #623

aschemmel-tech commented on docs/process/process_areas/verification/workproducts.rst:
```diff
@@ -0,0 +1,143 @@
+   :status: valid
+   :complies: std_wp__iso26262__software_13, std_wp__iso26262__support_12
+
+   Integration Testing verifies component architecture:
```
missing testing of Detailed Design (i.e. integration od units into component)

aschemmel-tech commented on docs/process/process_areas/verification/workproducts.rst:
```diff
@@ -0,0 +1,143 @@
+   :status: valid
+   :complies: std_wp__iso26262__software_11, std_wp__iso26262__support_12
+
+   Unit testing verifies component requirements and detailed design (traced to). Tooling defined
```
need to remove component requirements

aschemmel-tech commented on docs/process/process_areas/verification/workproducts.rst:
```diff
@@ -0,0 +1,143 @@
+
+As part of tool qualification as supporting function it is handled as follows
+
+     * :need:`rl__infrastructure_tooling_community` implementing :need:`wp__tool_eval`
```
recommend to put here link to "future process" or the plan document in PMP

aschemmel-tech commented on docs/process/process_areas/verification/index.rst:
```diff
@@ -0,0 +1,88 @@
+   * :need:`rl__contributor`
+   * :need:`rl__committer`
+
+Inspection activities on requirement, architecture and detailed design are handled within these process areas.
```
this is somehow a duplicate to what is in workproducts.rst recommend to delete one (maybe better the workproduct.rrst). And move this from "Roles" to "Workproducts" section.

pahmann replied:
I put it as introductory sentence into the verification_workproducts.rst

pahmann replied:
Whole roles parts is now put under verification_roles.rst and I will remove this part here.

pahmann commented on docs/process/process_areas/verification/index.rst:
```diff
@@ -0,0 +1,48 @@
+..
```
Now replicates the requirements index.rst adopted for verification.

masc2023 commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,188 @@
+     -
+   * - Description
+     - Text
+     - The description should include
```
Is this table correct, looks not?

hoe-jo commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,201 @@
+to other unit tests or component tests. This linking is done using metatags. This is also true for
+component integration tests linking to the component architecture.
+
+Comoponent tests are linked to component requirements directly.
```
typo: Comoponent

hoe-jo commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,201 @@
+General Traceability Concept
+============================
+
+To allow a traceability from any line of code to a written requirement, unit tests are linked
```
I would rathere state: Here only linkage to testcases and not to "any line of code"

hoe-jo commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,201 @@
+
+So if the content of the requirement is altered also the hash will change making it necessary to revisit the linkage of all test cases to the requirement again.
+
+If the status of the linked test case and the linkage document is valid the attribute *testcovered* shall be set to *YES* during the Sphinx Build.
```
Process Requirement?

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,105 @@
+test cases and reports the result in the same logging format as automated tests do.
+
+``[TODO: Reporting of manual execution results is pending]``
+
```
Maybe add how to test locally?

pahmann replied:
Added this finding to issue #623

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,105 @@
+
+Any failing test case requires an ISSUE.
+The passing rate of safety-critical test cases need to be 100% in order to release the affected component.
+In case of a lower pass rate than 100% for QM level tests, the :need:`rl__technical_lead` and
```
and?

pahmann replied:
the line was lost in limbo... Now it is added.

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_guidance.rst:
```diff
@@ -0,0 +1,105 @@
+-------------------------------
+
+Any failing test case requires an ISSUE.
+The passing rate of safety-critical test cases need to be 100% in order to release the affected component.
```
Maybe add a defined workflow for justification?

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+     - Type / Values
+     - Description
+     - Helpful links
+   * - Verifies
```
Shall we split verifies into two metatags: fully_verifies and partially_verifies

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+     - Links to one or more requirement Ids
+     -
+   * - Description
+     - Text
```
Somehow "Text" is missing in rendered document?
Maybe some list issue, since unordered lists are usually defined with "*" and not "-" in sphinx

pahmann replied:
For me it renders:
![image](https://github.com/user-attachments/assets/926eb41b-949b-4334-9658-d868803ff06b)

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+       - invalid
+     -
+     -
+   * - TestType
```
Maybe align with Concept (line 95ff)

pahmann replied:
Good point. I will make a reference to the concept to check existing types.

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_report_template.rst:
```diff
@@ -0,0 +1,124 @@
+      |
+
+      Note1: The verification report is valid for the module version tagged together with the report
+      Note2: All the above lists are generated automatically
```
Note2 is in one line with Note1 in the rendered document

hoe-jo commented on docs/process/process_areas/verification/guidance/verification_report_template.rst:
```diff
@@ -0,0 +1,124 @@
+      |
+
+      Note1: The verification report is valid for the platform version tagged together with the report
+      Note2: All the above lists are generated automatically
```
Note2 is in one line with Note1 in the rendered document

pahmann commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+
+   Test user login
+
+Additional information can also be found in :doc:`./verification_guidance`.
```
Update Link to need.

pahmann commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+
+Additional information can also be found in :doc:`./verification_guidance`.
+
+The specification is part of the test implementatin and has to comply to the requirements
```
typo

pahmann commented on docs/platform_management_plan/software_verification.rst:
```diff
@@ -12,16 +12,388 @@
+
+The traceability between verification relevant work products is one of the defined objectives.
+An overall overview of the different work products and their relationship is given in project
+context - see :doc:`/process/workproducts/index`.
```
check proper link.

pahmann commented on docs/process/process_areas/verification/workproducts.rst:
```diff
@@ -0,0 +1,153 @@
+The work products are handled within these process areas:
+
+     * :ref:`requirements_engineering` implementing :need:`wp__requirements__inspect`
+     * : ref : ` arch_design_process `
```
resolve missing refs.

pahmann commented on docs/platform_management_plan/software_verification.rst:
```diff
@@ -12,16 +12,423 @@
+Coverage of architectural design
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Beside :need:`wp__verification__comp_int_test` the following aspects define the coverage of the
```
Add feature architectural parts.

pahmann commented on docs/process/process_areas/verification/guidance/verification_specification.rst:
```diff
@@ -0,0 +1,137 @@
+       - Requirements-based test (Assumptions of Use are treated as requirements.)
+       - Design based tests
+
+       Consult the implementation of :need:`wp__verification__plan` for the full list of allowed types.
```
Put this in front of table as explanation. Or below, but not part of table.

pahmann commented:
@aschemmel-tech kindly do a review on the current state. I should have addressed all your pending concerns.

Missing part for final commit on this PR will be the transfer of TODOs into improvement issue(s).

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+     This may need involvment of :need:`rl__project_lead`,
+     :need:`rl__safety_manager`, and :need:`rl__security_manager`
+
+#. :need:`SW developer/Archticts <rl__contributor>`
```
typo Archticts -> Architects

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+   * Defining releae content
+   * Judge the releasable software state in case of failing test cases.
+
+     This may need involvment of :need:`rl__project_lead`,
```
typo involment -> involvement

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+
+#. :need:`Infrastructure/Tooling Developer <rl__infrastructure_tooling_community>`
+
+   * Enalbes execution of test cases in CI
```
typo Enalbes -> Enables

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+
+#. :need:`External Auditor <rl__external_auditor>`
+
+   * understand activities, planning, processes definition, and execution reports for verifcation activities
```
Start with Understand, as above

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+
+#. "Distributor" (external role)
+
+   * reexecution of test cases
```
Start with Reexecution as above

masc2023 commented on docs/process/process_areas/verification/verification_concept.rst:
```diff
@@ -0,0 +1,205 @@
+
+Following aspect should be considered when developing test cases:
+
+* **Comprehensive Coverage:** Test cases should cover all functional and non-functional requirements, including
```
do we really cover non-functional requirements, like process relevant topics?

aschemmel-tech replied:
yes, these are also ones from external compoents AoUs

pahmann commented:
@aschemmel-tech, from my perspective we are ready to merge. 🎉

aschemmel-tech commented on docs/platform_management_plan/software_verification.rst:
```diff
@@ -14,14 +14,372 @@
+As part of supporting activities:
+
+#. :need:`wp__verification__plan`
+#. :need:`wp__verification__specification`
```
agreed in working meeting Mar-10 to remove

aschemmel-tech reviewed (approved):
Agree, let's merge
